# scRNA_seq
A snakemake pipeline that can be used to analyze scRNA-seq data generated by 10x genomics. This pipeline can be used to analyze scRNA-seq, scVDJ-seq, and scCITE-seq. This pipeline works best when run on an lsf cluster.

Writen by Kristen Wells

To use:

1. Download and install miniconda3: For Linux
```{bash}
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh bash Miniconda3-latest-Linux-x86_64.sh
```
2. Install Snakemake:
```{bash}
conda install snakemake -c bioconda -c conda-forge
```

3. Update the config file (config.yaml) 
>* RAW_DATA: Location of the raw data 
>* IMAGES: Location of the image files (must be named the same as the raw data)
>* SAMPLES: the list of samples you want to test. This is the name that will be in the output files. This is a nested list. Format should be:
```
SAMPLES:
    {Sample1_name}:
        - {slide number (ex. "V10d07-047")}
        - {position on slide (A, B, C, or D)}
    {Sample2_name}:
        - {slide number (ex. "V10d07-047")}
        - {position on slide (A, B, C, or D)}
```
>* RESULTS: Path to the output directory
>* TRANSCRIPTOME: Path to the cellranger transcriptome reference
>* PROBE_SET: Path to cellranger probe set. Only fill in if sample is FFPE. Download the human here https://support.10xgenomics.com/spatial-gene-expression-ffpe/probe-sets/Visium-Human-Transcriptome-Probe-Set-v1-0-GRCh38-2020-A. Download the mouse here https://support.10xgenomics.com/spatial-gene-expression-ffpe/probe-sets/Visium-Mouse-Transcriptome-Probe-Set-v1-0-mm10-2020-A.
>* SLIDE_PATH: path to slide files. Download slide files here https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/using/count.
>* LSF_TEMPLATE: Path to an LSF template. One is included in this git repo.

4. Update snakecharmer.sh to your specific cluster specs. 
>* change the -q argument to the queue you want to use 

5. submit the job using `bsub < snakecharmer.sh`